### ISSUE TITLE
LDLT reports success but gives wrong answer, LU gives the correct answer

### ISSUE THREAD

**Issue by KristofferC** [DrTimothyAldenDavis/SuiteSparse#271]
This is a report based on an issue I encountered in Julia which uses SuiteSparse as the "backend" for the sparse solvers (https://github.com/JuliaSparse/SparseArrays.jl/issues/323). I don't have a pure C-repro of the issue yet because it would take some effort to do so and I am not sure if this issue is already known or maybe even invalid.

The matrix is real + symmetric + indefinite and quite small (306x306). Julia's poly algorithm for `\` for Hermitian matrices can be found at https://github.com/JuliaSparse/SparseArrays.jl/blob/a3116b95add064055c1f737abd98b5a5839c4147/src/solvers/cholmod.jl#L1575-L1587. It does a Cholesky factorization, if that fails, it tries to do an LDLT and if that fails it falls back to LU.

For this particular matrix the LDLT factorization reports success and this is used to solve the system.
However, the solution from this gives a residual which shows the system is not solved to any reasonable accuracy. Converting the matrix to dense (which would then use a lapack solver) or explicitly using an LU factorization from SuiteSparse gives accurate solutions.

The issue can be reprod (in Julia) by copy pasting the data in https://gist.githubusercontent.com/KristofferC/059b5576f718bb7de599abe04bdf6cf5/raw/20a98a25a7807c3f38352fd8ff1fa83bbc6f067d/K which defines the `colptr`, `rowval` and `nzval` vectors. Then running:

```julia
using SparseArrays, LinearAlgebra
K = SparseMatrixCSC(306, 306, colptr, rowval, nzval);
f = ones(size(K, 1));

u = K \ f;
u_lu = lu(K) \ f;
u_dense = Matrix(K) \ f;

compute_residual(K, u, f) = norm(f - K * u) / norm(f);

julia> @show compute_residual(K, u, f)
# 314.5598350582391

julia> @show compute_residual(K, u_lu, f)
# 1.9400817552325415e-12

julia> @show compute_residual(K, u_dense, f)
# 3.31677221636043e-13
```

The values in comments are the residual when the solution is computed with `\` (which ends up calling LDLT), `lu` and a dense solver, respectively.

I found this issue when I did some minor changes in a code which changed some of the very small values in the matrix. That caused `LDLT` to suddenly report success and the solution returned to be bogus.

I am happy to do further research, give further info if that would be helpful to diagnose this.

 - OS: macOS 13.1 M1
 - compiler [e.g. gcc 7.4, icc 19.0]: gcc
 - BLAS and LAPACK library. OpenBLAS
 - Version: 5.10.1

**Comment by DrTimothyAldenDavis** [comment]
I don't have a true LDL' solver.  I can compute and LDL' factorization only for a diagonal D with nonzero diagonal entries, and it's not suitable for all symmetric indefinite matrices.  In particular, for my simple LDL' solver inside CHOLMOD, and also my LDL package, the matrix must have the property that all leading submatrices are well conditioned.  I only have it because it's the format required for the update/downdate in CHOLMOD.

If Julia wants to use my LDL' factorization directly, it should be checking the condition number estimate.  It doesn't look like it's doing that.

**Comment by KristofferC** [comment]
Thanks for the speedy reply. I uploaded the matrix in matrix market format (I had to zip it due to file extension limitations by github): [K.mtx.zip](https://github.com/DrTimothyAldenDavis/SuiteSparse/files/10379961/K.mtx.zip)

Indeed, as written in https://github.com/JuliaSparse/SparseArrays.jl/issues/323#issuecomment-1376633972, it is very possible that the Julia is not using the API correctly.

To determine a "success" of a factorization, Julia looks at:

https://github.com/JuliaSparse/SparseArrays.jl/blob/43b4d01a7177a3469bd73a03d13c1bec9af32929/src/solvers/cholmod.jl#L1631-L1634

to check if the `minor` field in the `cholmod_factor` object is the same as the size of the matrix.

> If Julia wants to use my LDL' factorization directly, it should be checking the condition number estimate. It doesn't look like it's doing that.

Indeed, I don't see any checks for a condition number estimate so that might indeed be missing.
